pkg/nsx/services/builder.go

```diff
@@ -0,0 +1,1351 @@
+	var ruleAppliedGroupName string
+	appliedTo := rule.AppliedTo
+	ruleAppliedGroupID := fmt.Sprintf("sp_%s_%d_scope", obj.UID, idx)
+	if len(rule.Name) > 0 {
+		ruleAppliedGroupName = fmt.Sprintf("%s-scope", rule.Name)
+	} else {
+		ruleAppliedGroupName = fmt.Sprintf("%s-%d-scope", obj.ObjectMeta.Name, idx)
+	}
+	targetTags := service.buildTargetTags(obj, &appliedTo, idx)
+	ruleAppliedGroupPath := fmt.Sprintf("/infra/domains/%s/groups/%s", getDomain(service), ruleAppliedGroupID)
+	ruleAppliedGroup := model.Group{
+		Id:          &ruleAppliedGroupID,
+		DisplayName: &ruleAppliedGroupName,
+		Tags:        targetTags,
+	}
+
+	ruleGroupCount, ruleGroupTotalExprCount := 0, 0
+	criteriaCount, totalExprCount := 0, 0
+	var err error = nil
+	var errorMsg = ""
+	for i, target := range appliedTo {
+		criteriaCount, totalExprCount, err = service.updateTargetExpressions(obj, &target, &ruleAppliedGroup, i)
+		if err == nil {
+			ruleGroupCount += criteriaCount
+			ruleGroupTotalExprCount += totalExprCount
+		} else {
+			return nil, "", err
+		}
+	}
+	log.V(1).Info("build rule applied group criteria", "total criteria", ruleGroupCount, "total expressions of criteria", ruleGroupTotalExprCount)
+
+	if ruleGroupCount > MaxCriteria {
+		errorMsg = fmt.Sprintf("total counts of rule applied group criteria %d exceed NSX limit of %d", ruleGroupCount, MaxCriteria)
+	} else if ruleGroupTotalExprCount > MaxTotalCriteriaExpressions {
+		errorMsg = fmt.Sprintf("total expression counts in rule applied group criteria %d exceed NSX limit of %d", ruleGroupTotalExprCount, MaxTotalCriteriaExpressions)
+	}
+
+	if len(errorMsg) != 0 {
+		err = errors.New(errorMsg)
+		log.Error(err, "validate rule applied group criteria nsx limit failed")
+		return nil, "", err
+	}
+
+	return &ruleAppliedGroup, ruleAppliedGroupPath, nil
+}
+
+func (service *SecurityPolicyService) buildRuleID(obj *v1alpha1.SecurityPolicy, idx int) string {
+	return fmt.Sprintf("sp_%s_%d", obj.UID, idx)
+}
+
+func (service *SecurityPolicyService) buildRuleName(obj *v1alpha1.SecurityPolicy,
+	rule *v1alpha1.SecurityPolicyRule, idx int) string {
+	if len(rule.Name) > 0 {
+		return rule.Name
+	} else {
+		return fmt.Sprintf("%s-%d", obj.ObjectMeta.Name, idx)
+	}
+}
+
+func (service *SecurityPolicyService) buildRuleAppliedGroupByPolicy(
+	obj *v1alpha1.SecurityPolicy, nsxRuleSrcGroupPath string,
+	nsxRuleDstGroupPath string) (string, error) {
+	var nsxRuleAppliedGroupPath string
+	if len(obj.Spec.AppliedTo) == 0 {
+		return "", errors.New("appliedTo needs to be set in either spec or rules")
+	}
+	if nsxRuleSrcGroupPath == "ANY" && nsxRuleDstGroupPath == "ANY" {
+		// NSX-T manager will report error if all the rule's scope/src/dst are "ANY".
+		// So if the rule's scope is empty while policy's not, the rule's scope also
+		// will be set to the policy's scope to avoid this case.
```

Comment:
I saw this comment twice in this file. But in the previous code, I remember I only wrote it once.